### PR TITLE
Add Multi-Language Support

### PR DESCRIPTION
**Pull request by channarit1994**
Server Side's Multi-Language Support. (If you want multi-language show in-game. Edit Client too)

### CODE REVIEW

File: src/main/java/tools/data/input/GenericLittleEndianAccessor.java [hunk 1]
```diff
@@ -129,11 +130,17 @@ public double readDouble() {
      * @return The string read.
      */
     public final String readAsciiString(int n) {
-        char[] ret = new char[n];
+        byte ret[] = new byte[n];
```
Comment (P0nk) [inline src/main/java/tools/data/input/GenericLittleEndianAccessor.java:133]: Use `byte[] ret`, not `byte ret[]`

File: src/main/java/tools/data/input/GenericLittleEndianAccessor.java [hunk 2]
```diff
@@ -129,11 +130,17 @@ public double readDouble() {
...
         for (int x = 0; x < n; x++) {
-            ret[x] = (char) readByte();
+            ret[x] = (byte) readByte();
```
Comment (P0nk) [inline src/main/java/tools/data/input/GenericLittleEndianAccessor.java:135]: Redundant cast to byte. Remove `(byte)`.

File: src/main/java/tools/data/input/GenericLittleEndianAccessor.java [hunk 3]
```diff
@@ -129,11 +130,17 @@ public double readDouble() {
...
         }
-        return String.valueOf(ret);
+        try {
+            String str= new String(ret,CharsetConstants.MapleEncoding); // Multi-Language Support by MiLin
+            return str;
```
Comment (P0nk) [inline src/main/java/tools/data/input/GenericLittleEndianAccessor.java:139]: Inline the String. Do `return new String(ret, CharsetConstants.MapleEncoding);`

File: src/main/java/constants/string/CharsetConstants.java [hunk 4]
```diff
@@ -16,10 +18,13 @@
 public class CharsetConstants {
    
     public static MapleLanguageType MAPLE_TYPE = MapleLanguageType.LANGUAGE_US;
+    
+    public static Charset MapleEncoding = Charset.forName(MapleLanguageType.LANGUAGE_TH.getAscii()); // Multi-Language Support by MiLin (E.g. TH)
```
[the remark this answers is not in the document]
Comment (P0nk) [reply src/main/java/constants/string/CharsetConstants.java:22]: This field is not supposed to change, right? Make it `final`.

File: src/main/java/tools/data/output/GenericLittleEndianWriter.java [hunk 5]
```diff
@@ -131,7 +131,7 @@ public void writeAsciiString(String s) {
      */
     @Override
     public void writeMapleAsciiString(String s) {
-        writeShort((short) s.length());
+        writeShort((short) s.getBytes().length);
```
Comment (P0nk) [inline src/main/java/tools/data/output/GenericLittleEndianWriter.java:134]: This seems strange to me.

File: src/main/java/tools/StringUtil.java [hunk 6]
```diff
@@ -33,7 +33,7 @@
 	 */
 	public static String getLeftPaddedStr(String in, char padchar, int length) {
 		StringBuilder builder = new StringBuilder(length);
-		for (int x = in.length(); x < length; x++) {
+		for (int x = in.getBytes().length; x < length; x++) { // Multi-Language Support by MiLin
```
Comment (P0nk) [inline src/main/java/tools/StringUtil.java:36]: This seems strange to me.

File: src/main/java/tools/StringUtil.java [hunk 7]
```diff
@@ -51,7 +51,7 @@ public static String getLeftPaddedStr(String in, char padchar, int length) {
 	 */
 	public static String getRightPaddedStr(String in, char padchar, int length) {
 		StringBuilder builder = new StringBuilder(in);
-		for (int x = in.length(); x < length; x++) {
+		for (int x = in.getBytes().length; x < length; x++) { // Multi-Language Support by MiLin
```
Comment (P0nk) [inline src/main/java/tools/StringUtil.java:54]: This seems strange to me.

**Review by P0nk — CHANGES_REQUESTED** [review]
Thanks for the contribution. Commented on some simple style fixes. 

I will do some experimenting on my own to test the parts I found strange, and then get back to you with another verdict.

File: src/main/java/constants/string/CharsetConstants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (channarit1994) [reply src/main/java/constants/string/CharsetConstants.java]: I already change it please find a new commit.

**Comment by P0nk** [comment]
I will put this on hold for a while, and get back to it later. I definitely like the idea, it will just have to wait.

**Comment by channarit1994** [comment]
> I will put this on hold for a while, and get back to it later. I definitely like the idea, it will just have to wait.

Thank you

**Comment by P0nk** [comment]
Implemented in a separate PR: https://github.com/P0nk/Cosmic/pull/51

**Pull request closed without merging**